Doc/library/unittest.mock.rst

```diff
@@ -1078,6 +1080,37 @@ object::
       [call('foo'), call('bar')]
 
 
+.. class:: ThreadingMock(spec=None, side_effect=None, return_value=DEFAULT, wraps=None, name=None, spec_set=None, unsafe=False, **kwargs)
+
+  A version of :class:`MagicMock` for multithreading tests. The
+  :class:`ThreadingMock` object provides extra methods to wait for a call to
+  happen on a different thread, rather than assert on it immediately.
+
+  .. method:: wait_until_called(mock_timeout=None)
+
+      Waits until the the mock is called.
+      If ``mock_timeout`` is set, after that number of seconds waiting,
+      it raises an :exc:`AssertionError`, waits forever otherwise.
+
+        >>> mock = ThreadingMock()
+        >>> thread = threading.Thread(target=mock)
+        >>> thread.start()
+        >>> mock.wait_until_called(mock_timeout=1)
+        >>> thread.join()
+
+  .. method:: wait_until_any_call(*args, mock_timeout=None, **kwargs)
```

Comment:
I would prefer to put the timeout as the first parameter, maybe as a positional-only parameter to avoid any risk of conflict with **kwargs. If you want to block, pass explicitly None. If having to pass None bothers you, maybe add a second function which always block?

Reply:
We tried to namespace it as it is done with other arguments in `mock` to avoid conflicts indeed.

I did not think about potentially having 2 different methods though, that might be nice indeed.

Something like:

`wait_until_any_call` and `wait_until_any_call_with_timeout`?

Thoughts @lisroach @voidspace @tirkarthi ?

Reply:
Usually timeout=None is called "blocking" in Python, "timeout=0.0" is called "non-blocking" and "timeout=value" has no clear name... If we choose to have 2 methods, maybe "blocking" could be used for the timeout=None case: block_until_called(...) vs wait_until_called(timeout, ...).

I'm not strongly opposed to mock_timeout=value parameter.

Is it a good practice to use a wait without upper bound in an unit test? For example, is it fine if your test completes in 1 minute? 1 hour? 1 day? 1 week?

My experience with the Python test suite is a that a timeout helps to get an output when a test hangs.

Reply:
I am slightly more inclined to just having `mock_timeout`, I think it is the simplest and easiest to discover, the methods are similar to the existing ones and across the two of them.
The "only" downside is that they might collide with an existing argument, but we are namespacing it already, it is `mock_timeout`, not `timeout`. Having [a quick search in github](https://github.com/search?l=Python&q=mock_timeout+NOT+test&type=Code) seems to indicate it might not be an issue. Additionally, if users wanted to fake a function that had an argument `mock_timeout` they can do it so by using `side_effect` (basically not being able to use this function).

In terms of having a default timeout, I agree with [your previous comment here](https://github.com/python/cpython/pull/12818#discussion_r275276284) which prefers to not have a default at all. Any default is trying to guess what the user is mocking, which can be result in flaky tests. Say you are working with a function that takes 1 second and we set the 1 second as a default for example.
The "only harm" with blocking forever is that the test suite will not finish and it will be eventually terminated by something else (user, CI, pytest, etc.). I think it is much better to have a default that blocks forever (so there are no surprises for the user based on the environment for example) and let them specify a timeout if they really want, as they will know what is the right value.

Reply:
As a note, IIRC when speaking with @voidspace, he liked the idea of having the timeout namespaced. He can further comment.

Reply:
I'm a little confused: can you give an example where naming the parameter `timeout` would conflict?

Reply:
Say you want to mock the following function:

```python
def my_cool_socket_func(timeout=3):
    pass
```

You then wait for it to be called as:
```
m = ThreadingMock(my_cool_socket_func)
# some other thread runs: m(timeout=4)
m.wait_until_any_call(timeout=4)
```

The user might expect the `wait_until_any_call` to pass, but it will fail after 4 seconds, as the timeout keyword matches our waiting timeout, not the parameter timeout. Not a huge concern though.

We could split the function as Victor was suggesting and just rename it to timeout if you prefer, having:

```
def block_until_any_call(*args, **kwargs):
    pass
def wait_until_any_call(timeout, /, *args, **kwargs):
    pass
```

This way:
- Block function is kept clean and just takes the arguments as the function takes them when it is called.
- Wait function always takes timeout as the first argument and as it is possitional only, it cannot collide with a function parameter "timeout".

In short:

```
m = ThreadingMock(my_cool_socket_func)
# some other thread runs: m(timeout=4)
m.wait_until_any_call(timeout=4)  # Fails, missing positional arg timeout
m.block_until_any_call(timeout=4)  # Works, blocks until it is called
m.wait_until_any_call(4)  # Works, blocks for 4 seconds
```

Reply:
The trouble with positional-only stuff is it makes backporting really hard :-( 
(I'm also not sure it doesn't cause other problems of the a similar type...)

How about having `waitaible_for_call(*args, **kw)` that returns an object that has a `wait(timeout=None)` method?
(a bit like `mock_calls`, and with maybe `waitable()` that return an object that waits for the next call of any type?)

Reply:
> The trouble with positional-only stuff is it makes backporting really hard :-(

Have you considered drooping support for Python < 3.7?
3.7 will be marked EOL this very month.

Reply:
The backport still supports 3.6!
Unfortunately, cpython's much shortened support windows don't work for everyone...